1. 

(X_train=80%)
(X_test=20%)

A partir del archivo "iris.csv" que se corresponde con datos relacionados con tres tipos de plantas (el target es el tipo de planta):

- Mediante regresión logística, predice qué planta se corresponde con los ejemplos: [[15.7,2.8,9.5,0.1]] y [[1.7,3.8,4.5,11.3]]
- ¿Qué acierto tiene tu modelo? (score)
- Utiliza un conjunto de test con el 40% de los datos: ¿cambian los resultados anteriores? ¿por qué?
- Elimina la clase "Virginica" de tus datos. Repite el proceso anterior, ¿cambian los resultados?
- ¿Consideras la regresión logística un buen algoritmo para estos datos? ¿por qué?

In [13]:
# Libraries
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix

In [3]:
# Loading data
filename = '../data/iris.csv'
df = pd.read_csv(filename)
df

sepal.length  sepal.width  petal.length  petal.width    variety
0             5.1          3.5           1.4          0.2     Setosa
1             4.9          3.0           1.4          0.2     Setosa
2             4.7          3.2           1.3          0.2     Setosa
3             4.6          3.1           1.5          0.2     Setosa
4             5.0          3.6           1.4          0.2     Setosa
..            ...          ...           ...          ...        ...
145           6.7          3.0           5.2          2.3  Virginica
146           6.3          2.5           5.0          1.9  Virginica
147           6.5          3.0           5.2          2.0  Virginica
148           6.2          3.4           5.4          2.3  Virginica
149           5.9          3.0           5.1          1.8  Virginica

[150 rows x 5 columns]

In [4]:
# Creating X and y variables
X = np.array(df.iloc[:, :-1])
y = np.array(df['variety'])
print('X shape:', X.shape)
print('y shape:', y.shape)

X shape: (150, 4)
y shape: (150,)


In [5]:
# Label Encoder transform allows us to map string groups to number classes
le = preprocessing.LabelEncoder()
y = le.fit_transform(y)
y
# We have transformed the varieties into number groups: [0 - Setosa, 1 - Versicolor, 2 - Virginica] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [6]:
# Split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=17)
print(X_train.shape)
print(X_test.shape)

(120, 4)
(30, 4)


In [9]:
# Creating a Logistic Regression model 
model = LogisticRegression()
model.fit(X_train, y_train)

LogisticRegression()

In [10]:
# Predicting our test data
predictions = model.predict(X_test)
print(predictions)

[0 1 2 1 2 2 1 2 1 2 2 0 1 0 2 0 0 2 2 2 1 0 2 1 1 1 1 1 0 1]


In [14]:
# Confusion Matrix
confusion_matrix(y_test, predictions)

array([[ 7,  0,  0],
       [ 0, 11,  0],
       [ 0,  1, 11]])

In [15]:
# Applying the model to predict the examples
# Example 1
model.predict([[15.7,2.8,9.5,0.1]])
# Prediction: Virginica

array([2])

In [16]:
# Example 2
model.predict([[1.7,3.8,4.5,11.3]])
# Prediction: Virginica

array([2])

In [17]:
# Score of our prediction
model.score(X_train, y_train)

0.975

# Repeating the process with 40 train 60 test

In [21]:
# Split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.60, random_state=17)
print(X_train.shape)
print(X_test.shape)
# Creating a Logistic Regression model 
model = LogisticRegression()
model.fit(X_train, y_train)

(60, 4)
(90, 4)


LogisticRegression()

In [22]:
# Applying the model to predict the examples
# Example 1
model.predict([[15.7,2.8,9.5,0.1]])
# Prediction: Virginica

array([2])

In [23]:
# Example 2
model.predict([[1.7,3.8,4.5,11.3]])
# Prediction: Virginica

array([2])

In [24]:
# Score of our prediction
model.score(X_train, y_train)

0.9666666666666667

Using a 60% test size, the predictions for the examples were the same, but the model did drop in score from 97.5% to 96.6%. The lack of training data can cause the model not to train as optimally as it can

# Removing Virginica


In [28]:
# Creating X and y variables
X = np.array(df[df['variety'] != 'Virginica'].iloc[:, :-1])
y = np.array(df[df['variety'] != 'Virginica']['variety'])
print('X shape:', X.shape)
print('y shape:', y.shape)

X shape: (100, 4)
y shape: (100,)


In [29]:
# Label Encoder transform allows us to map string groups to number classes
le = preprocessing.LabelEncoder()
y = le.fit_transform(y)
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1])

In [30]:
# Split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=17)
print(X_train.shape)
print(X_test.shape)

(80, 4)
(20, 4)


In [31]:
# Creating a Logistic Regression model 
model = LogisticRegression()
model.fit(X_train, y_train)

LogisticRegression()

In [32]:
# Applying the model to predict the examples
# Example 1
model.predict([[15.7,2.8,9.5,0.1]])
# Prediction: Versicolor

array([1])

In [33]:
# Example 2
model.predict([[1.7,3.8,4.5,11.3]])
# Prediction: Versicolor

array([1])

In [36]:
# Score of our prediction
model.score(X_train, y_train)

1.0

Removing the Virginica variety we obtained a score of the model of 1, which is perfect and the examples were classified as Versicolor, because Virginica was not an option. When classification is binary (2 groups), problems tend to be simplified and easier as a result, in most cases.

# Verdict: Is logistic regression a good algorithm for these types of data?


I would say yes, because the model takes into account the different numerical values of the x data and how they relate to the target values for a precise classification algorithm